eio/io.ml

```diff
@@ -0,0 +1,181 @@
+open Websocket
+open Astring
+open Eio
+
+type mode = Client of (int -> string) | Server
+
+let is_client mode = mode <> Server
+
+let xor mask msg =
+  for i = 0 to Bytes.length msg - 1 do
+    (* masking msg to send *)
+    Bytes.set msg i
+      Char.(to_int mask.[i mod 4] lxor to_int (Bytes.get msg i) |> of_byte)
+  done
+
+let is_bit_set idx v = (v lsr idx) land 1 = 1
+let set_bit v idx b = if b then v lor (1 lsl idx) else v land lnot (1 lsl idx)
+let int_value shift len v = (v lsr shift) land ((1 lsl len) - 1)
+
+(* Eio only uses Cstructs so we lose 0copy :(! *)
+let read_exactly src remaining =
+  let open Cohttp_eio.Server in
+  try
+    Buf_read.ensure src remaining ;
+    let t = Some (Cstruct.to_string (Buf_read.peek src) ~off:0 ~len:remaining) in
+    Buf_read.consume src remaining ;
+    t
+  with End_of_file -> None
+
+let read_uint16 ic =
+  match read_exactly ic 2 with
+  | None -> None
+  | Some s -> Some (EndianString.BigEndian.get_uint16 s 0)
+
+let read_int64 ic =
+  match read_exactly ic 8 with
+  | None -> None
+  | Some s -> Some (Int64.to_int @@ EndianString.BigEndian.get_int64 s 0)
+
+let write_frame_to_buf ~mode buf fr =
+  let scratch = Bytes.create 8 in
+  let open Frame in
+  let content = Bytes.unsafe_of_string fr.content in
+  let len = Bytes.length content in
+  let opcode = Opcode.to_enum fr.opcode in
+  let payload_len =
+    match len with
+    | n when n < 126 -> len
+    | n when n < 1 lsl 16 -> 126
+    | _ -> 127 in
+  let hdr = set_bit 0 15 fr.final in
+  (* We do not support extensions for now *)
+  let hdr = hdr lor (opcode lsl 8) in
+  let hdr = set_bit hdr 7 (is_client mode) in
+  let hdr = hdr lor payload_len in
+  (* Payload len is guaranteed to fit in 7 bits *)
+  EndianBytes.BigEndian.set_int16 scratch 0 hdr ;
+  Buf_write.bytes ~off:0 ~len:2 buf scratch ;
```

Comment:
You can use `Buf_write.BE.uint16` here. You shouldn't need your own scratch buffer.

Reply:
Yep, thank you